bionic/deriver.py

```diff
@@ -147,41 +184,112 @@ def _register_static_providers(self):
+                # converted value can still be different from the draft value.
+                if isinstance(child_dnode, ast.EntityNode):
+                    register_provider(PassthroughProvider(child_dnode, dnode))
+                    register_followup(child_dnode, dnode)
+
+                # If we have a draft of a tuple (like "<D1, D2>"), we "distribute" the
+                # draft-ness over the childern (producing "<D1>, <D2>"). The idea is
+                # that "<D1, D2>" is an unnormalized tuple containing unnormalized
+                # values, while "<D1>, <D2>" is a normalized tuple containing
+                # unnormalized values. (In the case of tuples, the normalization
+                # doesn't actually do anything, but in the future it might; for
+                # example, we might allow the unnormalized value to be a list, and then
+                # normalize it by converting it to an actual tuple. TODO Let's actually
+                # do this!)
+                elif isinstance(child_dnode, ast.TupleNode):
+                    out_dnode = ast.TupleNode(
+                        ast.DraftNode(grandchild_dnode)
+                        for grandchild_dnode in child_dnode.children
+                    )
+                    register_provider(PassthroughProvider(out_dnode, dnode))
+                    register_followup(out_dnode, dnode)
+
+                else:
+                    raise AssertionError(
+                        f"Unexpected dnode type {type(dnode)!r} for dnode {dnode!r}"
+                    )
+
+            # If this provider generates a tuple value ("D1, D2"), we will add
+            # providers for extracting all the components ("D1" and "D2"). Note that
+            # the components must be drafts, because the only situation where we
+            # register a tuple provider is the "distribute" step above. (Remember that
+            # if a user-provided function outputs a tuple, it will be wrapped in a
+            # draft node and get handled by the case above first.)
+            elif isinstance(dnode, ast.TupleNode):
                 for child_dnode in dnode.children:
-                    child_provider = TupleDeconstructionProvider(child_dnode, dnode)
-                    register_provider(child_provider)
+                    assert isinstance(child_dnode, ast.DraftNode)
+                    register_provider(TupleDeconstructionProvider(child_dnode, dnode))
+                    register_followup(child_dnode, dnode)
+
+            else:
+                # The only remaining node type is an entity node, and there's no
+                # additional work for us to do on those.
+                assert isinstance(dnode, ast.EntityNode)
 
         for provider in set(self._flow_state.providers_by_name.values()):
             register_provider(provider)
 
         self._static_providers_by_dnode = static_providers_by_dnode
+        self._followup_dnode_lists_by_dnode = dict(followup_dnode_lists_by_dnode)
```

Comment:
Why do you need to cast this to a dict again?

Reply:
I just wanted to avoid any possible surprising behavior downstream, since other parts of the code may not realize that this is a `defaultdict`, not a `dict`.  (I don't anticipate any specific problems here, since the object's behavior will make sense either way, but it still seems safer to me.) I can add a comment.